csharp/src/Google.Protobuf/ExtensionSet.cs

```diff
@@ -63,7 +64,31 @@ public static class ExtensionSet
             IExtensionValue value;
             if (TryGetValue(ref set, extension, out value))
             {
-                return ((ExtensionValue<TValue>)value).GetValue();
+                // The stored ExtensionValue can be a different type to what is being requested.
+                // This happens when the same extension proto is compiled in different assemblies.
+                // To allow consuming assemblies to still get the value when the TValue type is
+                // different, this get method:
+                // 1. Attempts to cast the value to the expected ExtensionValue<TValue>.
+                //    This is the usual case. It is used first because it avoids possibly boxing the value.
+                // 2. Fallback to get the value as object from IExtensionValue then casting.
+                //    This allows for someone to specify a TValue of object. They can then convert
+                //    the values to bytes and reparse using expected value.
+                // 3. If neither of these work, throw a user friendly error that the types aren't compatible.
+                if (value is ExtensionValue<TValue> extensionValue)
+                {
+                    return extensionValue.GetValue();
+                }
+                else if (value.GetValue() is TValue underlyingValue)
+                {
+                    return underlyingValue;
+                }
+                else
+                {
+                    var storedType = value.GetType().GetTypeInfo().GenericTypeArguments[0];

```

Comment:
I think it would be worth leaving a comment here explaining why we believe all `IExtensionValue` values will have generic type arguments - and add a comment to `IExtensionValue` to advise that anyone creating a non-generic `IExtensionValue` implementation should change this code. Alternatively, make this code more defensive.

Reply:
Made it more defensive.